### ISSUE TITLE
LWG-2839 Self-move-assignment of library types, again

### ISSUE THREAD

**Issue by CaseyCarter** [microsoft/STL#1485]
LWG-2839 "Self-move-assignment of library types, again" clarifies that move-assigning most Standard Library types to themselves leaves the object in a valid-but-unspecified state. While we are fairly fastidious about handling self-move and self-swap correctly when authoring new types, I'm not sure that our neanderthal predecessors paid as much attention. We need to audit all of the move assignment operators in the STL to ensure that self-move leaves the class in a valid state except when the standard explicitly specifies otherwise.

**Comment by StephanTLavavej** [comment]
To make this more actionable, I've added a list of all of the C++23 headers, so we can audit them one-by-one. It may be convenient to use `_EXPORT_STD` to find the public classes defined by each header.

Because I'm a lazy kitty, I haven't yet bothered checking off the headers that obviously define no types (like `<version>`).

**Comment by AlexGuteniev** [comment]
Is this a good move-assignment implementation?
https://github.com/microsoft/STL/blob/5762e6bcaf7f5f8b5dba0a9aabf0acbd0e335e80/stl/inc/barrier#L55-L59

I'm not sure. On one hand it leaves arrival token in valid but unspecified state. On the other hand the token information is lost.

**Comment by StephanTLavavej** [comment]
Looks good enough to me. Valid-but-unspecified means exactly what it says on the box - it doesn't say anything about being a no-op.

**Comment by himisawww** [comment]
I see some strange (and seems intended) behavior when self-move-assigning `_Node_handle`:
https://github.com/microsoft/STL/blob/2e27f1f05a155bb3637dad04b4d9c4c4590c8542/stl/inc/xnode_handle.h#L128-L131

This, while following what [standard](https://eel.is/c++draft/container.node.cons#itemdecl:1) said, and does `leave the object in a valid-but-unspecified state`, but also clears the node and lost the data it holds when self-move-assigning.

This behavior is kind of surprising and does not align with most of other stl types do, they generally do no-op when self-move-assigning, like `std::string/vector/...`.

My questions are:
- `_Node_handle` clears itself when self-move-assigning. Is this behavior mandatory and cannot being just no-op because what the standard said about its implementation?
- Most of msvc-stl types is implemented doing no-op when self-move-assigning. Is this behavior not guaranteed by standard, not encouraged by standard, and should not be relied on when using them?

**Comment by frederick-vs-ja** [comment]
> * `_Node_handle` clears itself when self-move-assigning. Is this behavior mandatory and cannot being just no-op because what the standard said about its implementation?

Yes. This is "otherwise specified" in [[container.node.cons]/3.4](https://eel.is/c++draft/container.node.cons#3.4).

> * Most of msvc-stl types is implemented doing no-op when self-move-assigning. Is this behavior not guaranteed by standard, not encouraged by standard, and should not be relied on when using them?

A few of no-op treatments are standard-guaranteed, e.g. [[range.move.wrap]](https://eel.is/c++draft/range.move.wrap) and [[thread.jthread.cons]/12](https://eel.is/c++draft/thread.jthread.cons#12). The rest are not.

I don't think we should rely on the resulted value after self-move-assignment, even though sometimes no-op is considered to be the most reasonable choice and standardized. I think we can rely on the validity of the self-move-assignment operation itself and the result of subsequent (non-self) assignment to that object, though.
